README.md
# Logstash Plugin - public_ip

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

This Logstash filter plugin is used to determine ip version and if the ip address is public.

The plugin uses the following cidr ranges to determine if the ipaddress is in a non-public range.
```
# NON-PUBLIC REASONS
# ---------------
# 0.0.0.0/8 = RFC1700: reserved as a source address only
# 10.0.0.0/8 = RFC1918: reserved for private networking
# 100.64.0.0/10 = RFC6598: reserved for service provider shared address space but may be used in a manner similiar to RF1918
# 127.0.0.0/8 = RFC1112: assigned for use as the Internet host loopback address
# 169.254.0.0/16 = RFC3927: used for link-local addressing in Internet Protocol Version 4
# 172.16.0.0/12 = RFC1918: reserved for private networking
# 192.0.0.0/24 = RFC6890: IETF Protocol Assignments
# 192.0.0.8/32 = RFC7600: IPv4 dummy address
# 192.0.2.0/24 = RFC5737: Assigned as TEST-NET-1, documentation and examples
# 192.31.196.0/24 = RFC7535: AS112-v4
# 192.52.193.0/24 = RFC7450: AMT
# 192.88.99.0/24 = RFC7526: Reserved. Formerly used for IPv6 to IPv4 relay (included IPv6 address block 2002::/16
# 192.168.0.0/16 = RFC1918: reserved for private networking
# 198.18.0.0/15 = RFC2544: Used for benchmark testing of inter-network communications between two separate subnets
# 198.51.100.0/24 = RFC5737: Assigned as TEST-NET-2, documentation and examples
# 203.0.113.0/24 = RFC5737: Assigned as TEST-NET-3, documentation and examples
# 224.0.0.0/4 = RFC1112: In use for IP multicast (Former Class D network)
# 240.0.0.0/4 = RFC6890: Reserved for future use. (Former Class E network) 
# 255.255.255.255/32 = RFC8190: Reserved for the limited broadcast destination address
# fc00::/7 = RFC4193: Unique Local Address
# fe80::/10 = RFC4291: Link-Local Address
# ff00::/8 = RFC4291: Multicast Address
# 2001:db8::/32 = RFC3849: Addresses used in documentation and example source code
# 2001:20::/28 = RFC7343: Prefix for Overlay Routable Cryptographic Hash Identifiers Version 2
# ::1/128 = RFC8190: Loopback address to local host
# ::/128 = RFC8190: Unspecified address
# 100::/64 = RFC6666: Discard Prefix
# 64:ff9b::/96 = RFC6052: IPv4/IPv6 translation
```

## Installation
```
bin/logstash-plugin install logstash-filter-public_ip
```
- Start Logstash and proceed to test the plugin

#### USAGE
```
filter {
  public_ip {
    source => "src_ip"
    target_ipv => "src_ipv"
    target_pub_ip => "src_public_ip"
  }
}
```


CHANGELOG.md
## 1.0.2
 - fixed bad release of gem
## 1.0.1
  - removed ipvlookup class and placed code in main plugin
  - improved speed on lookup
## 1.0.0
  - changed array lookup
## 0.1.1
  - fixed bug in cidr logic
## 0.1.0
  - Plugin created with the logstash plugin generator


spec/filters/public_ip_spec.rb
# encoding: utf-8
require_relative '../spec_helper'
require "logstash/filters/public_ip"

describe LogStash::Filters::PublicIp do
  describe "ipversion" do
    let(:config) do <<-CONFIG
      filter {
        public_ip {
          source => "ip"
          target_ipv => "ipv"
          target_pub_ip => "pubip"
        }
      }
    CONFIG
    end
    
    sample("ip" => "10.0.0.1") do
      expect(subject.get("ipv")).to eq('4')
    end
  end

  describe "invalid ip" do
    let(:config) do <<-CONFIG
      filter {
        public_ip {
          source => "ip"
          target_ipv => "ipv"
          target_pub_ip => "pubip"
        }
      }
    CONFIG
    end

    sample("ip" => "10000.0.0.1") do
      expect(subject.get("tags")).to include("_invalid_ip")
    end
  end

  describe "ipv4 pubip false" do
    let(:config) do <<-CONFIG
      filter {
        public_ip {
          source => "ip"
          target_ipv => "ipv"
          target_pub_ip => "pubip"
        }
      }
    CONFIG
    end

    sample("ip" => "10.0.0.1") do
      expect(subject.get("pubip")).to eq(false)
    end
  end

  describe "ipv6 pubip true" do
    let(:config) do <<-CONFIG
      filter {
        public_ip {
          source => "src_ip"
          target_ipv => "[src][ipv]"
          target_pub_ip => "[src][pubip]"
        }
      }
    CONFIG
    end

    sample("src_ip" => "2606:4700:4700::1001") do
      expect(subject.get("[src][pubip]")).to eq(true)
    end
  end

  describe "no valid source" do
    let(:config) do <<-CONFIG
      filter {
        public_ip {
          source => "ip"
          target_ipv => "ipv"
          target_pub_ip => "pubip"
        }
      }
    CONFIG
    end

    sample("ip" => nil) do
      expect(subject.get("tags")).to include("_invalid_ip")
    end
  end

end


lib/logstash/filters/ipvlookup.rb
# encoding: utf-8
require 'ipaddr'

# NON-PUBLIC REASONS
# ---------------
# 0.0.0.0/8 = RFC1700: reserved as a source address only
# 10.0.0.0/8 = RFC1918: reserved for private networking
# 100.64.0.0/10 = RFC6598: reserved for service provider shared address space but may be used in a manner similiar to RF1918
# 127.0.0.0/8 = RFC1112: assigned for use as the Internet host loopback address
# 169.254.0.0/16 = RFC3927: used for link-local addressing in Internet Protocol Version 4
# 172.16.0.0/12 = RFC1918: reserved for private networking
# 192.0.0.0/24 = RFC6890: IETF Protocol Assignments
# 192.0.0.8/32 = RFC7600: IPv4 dummy address
# 192.0.2.0/24 = RFC5737: Assigned as TEST-NET-1, documentation and examples
# 192.31.196.0/24 = RFC7535: AS112-v4
# 192.52.193.0/24 = RFC7450: AMT
# 192.88.99.0/24 = RFC7526: Reserved. Formerly used for IPv6 to IPv4 relay (included IPv6 address block 2002::/16
# 192.168.0.0/16 = RFC1918: reserved for private networking
# 198.18.0.0/15 = RFC2544: Used for benchmark testing of inter-network communications between two separate subnets
# 198.51.100.0/24 = RFC5737: Assigned as TEST-NET-2, documentation and examples
# 203.0.113.0/24 = RFC5737: Assigned as TEST-NET-3, documentation and examples
# 224.0.0.0/4 = RFC1112: In use for IP multicast (Former Class D network)
# 240.0.0.0/4 = RFC6890: Reserved for future use. (Former Class E network) 
# 255.255.255.255/32 = RFC8190: Reserved for the limited broadcast destination address
# fc00::/7 = RFC4193: Unique Local Address
# fe80::/10 = RFC4291: Link-Local Address
# ff00::/8 = RFC4291: Multicast Address
# 2001:db8::/32 = RFC3849: Addresses used in documentation and example source code
# 2001:20::/28 = RFC7343: Prefix for Overlay Routable Cryptographic Hash Identifiers Version 2
# ::1/128 = RFC8190: Loopback address to local host
# ::/128 = RFC8190: Unspecified address
# 100::/64 = RFC6666: Discard Prefix
# 64:ff9b::/96 = RFC6052: IPv4/IPv6 translation

class Ipvlookup
  def initialize(ipaddress)
    @ipaddress = ipaddress
    @cidr = [
      "0.0.0.0/8",
      "10.0.0.0/8",
      "100.64.0.0/10",
      "127.0.0.0/8",
      "169.254.0.0/16",
      "172.16.0.0/12",
      "192.0.0.0/24",
      "192.0.0.8/32",
      "192.0.2.0/24",
      "192.31.196.0/24",
      "192.52.193.0/24",
      "192.88.99.0/24",
      "192.168.0.0/16",
      "198.18.0.0/15",
      "198.51.100.0/24",
      "203.0.113.0/24",
      "224.0.0.0/4",
      "240.0.0.0/4",
      "255.255.255.255/32",
      "fc00::/7",
      "fe80::/10",
      "ff00::/8",
      "2001:db8::/32",
      "2001:20::/28",
      "::1/128",
      "::/128",
      "100::/64",
      "64:ff9b::/96"
    ]
  end
  def validip
    begin
      @src = IPAddr.new("#{@ipaddress}")
      if @src.ipv4?
        @valid = true
        @ipv = '4'
      elsif @src.ipv6?
        @valid = true
        @ipv = '6'
      else
        @valid = false
      end
    rescue
      @valid = false
    end
    return @valid
  end
  def ipversion
    if @valid == true
      return "#{@ipv}"
    end
  end
  def pubip
    if @valid == true
      prvip = @cidr.collect{|p|
        cnet = IPAddr.new(p)
        cnet.include?(@src)
      }
      if prvip.include? true
        public_ip = false
      else
        public_ip = true
      end
      return public_ip
  
      #public_ip = true
      #for p in @cidr do
      #  cidr = IPAddr.new(p)
      #  if cidr.include?(@src) == true
      #    public_ip = false
      #  end
      #end
      #return public_ip
    end
  end
end


lib/logstash/filters/public_ip.rb
# encoding: utf-8
require "logstash/filters/base"
require "logstash/namespace"
require 'ipaddr'

# NON-PUBLIC CIDRS THIS PLUGIN USES TO VERIFY IF THE IP IS PUBLIC OR PRIVATE
# ---------------
# 0.0.0.0/8 = RFC1700: reserved as a source address only
# 10.0.0.0/8 = RFC1918: reserved for private networking
# 100.64.0.0/10 = RFC6598: reserved for service provider shared address space but may be used in a manner similiar to RF1918
# 127.0.0.0/8 = RFC1112: assigned for use as the Internet host loopback address
# 169.254.0.0/16 = RFC3927: used for link-local addressing in Internet Protocol Version 4
# 172.16.0.0/12 = RFC1918: reserved for private networking
# 192.0.0.0/24 = RFC6890: IETF Protocol Assignments
# 192.0.0.8/32 = RFC7600: IPv4 dummy address
# 192.0.2.0/24 = RFC5737: Assigned as TEST-NET-1, documentation and examples
# 192.31.196.0/24 = RFC7535: AS112-v4
# 192.52.193.0/24 = RFC7450: AMT
# 192.88.99.0/24 = RFC7526: Reserved. Formerly used for IPv6 to IPv4 relay (included IPv6 address block 2002::/16
# 192.168.0.0/16 = RFC1918: reserved for private networking
# 198.18.0.0/15 = RFC2544: Used for benchmark testing of inter-network communications between two separate subnets
# 198.51.100.0/24 = RFC5737: Assigned as TEST-NET-2, documentation and examples
# 203.0.113.0/24 = RFC5737: Assigned as TEST-NET-3, documentation and examples
# 224.0.0.0/4 = RFC1112: In use for IP multicast (Former Class D network)
# 240.0.0.0/4 = RFC6890: Reserved for future use. (Former Class E network) 
# 255.255.255.255/32 = RFC8190: Reserved for the limited broadcast destination address
# fc00::/7 = RFC4193: Unique Local Address
# fe80::/10 = RFC4291: Link-Local Address
# ff00::/8 = RFC4291: Multicast Address
# 2001:db8::/32 = RFC3849: Addresses used in documentation and example source code
# 2001:20::/28 = RFC7343: Prefix for Overlay Routable Cryptographic Hash Identifiers Version 2
# ::1/128 = RFC8190: Loopback address to local host
# ::/128 = RFC8190: Unspecified address
# 100::/64 = RFC6666: Discard Prefix
# 64:ff9b::/96 = RFC6052: IPv4/IPv6 translation

class LogStash::Filters::PublicIp < LogStash::Filters::Base

  # configure this filter from your Logstash config.
  #
  # filter {
  #   public_ip {
  #     source => "src_ip"
  #     target_ipv => "src_ipv"
  #     target_pub_ip => "src_public_ip"
  #   }
  # }
  #
  config_name "public_ip"

  # Field with source ip address
  config :source, :validate => :string, :required => true
  # target field to put ip vesrion data
  config :target_ipv, :validate => :string, :default => "ip_version"
  # target field to put ip vesrion data
  config :target_pub_ip, :validate => :string, :default => "public_ip"
  # tag invalid ip
  config :tag_on_invalid_ip, :validate => :array, :default => ["_invalid_ip"]

  public
  def register
    # NON-PUBLIC CIDR ARRAY
    @cidr = [
      "0.0.0.0/8",
      "10.0.0.0/8",
      "100.64.0.0/10",
      "127.0.0.0/8",
      "169.254.0.0/16",
      "172.16.0.0/12",
      "192.0.0.0/24",
      "192.0.0.8/32",
      "192.0.2.0/24",
      "192.31.196.0/24",
      "192.52.193.0/24",
      "192.88.99.0/24",
      "192.168.0.0/16",
      "198.18.0.0/15",
      "198.51.100.0/24",
      "203.0.113.0/24",
      "224.0.0.0/4",
      "240.0.0.0/4",
      "255.255.255.255/32",
      "fc00::/7",
      "fe80::/10",
      "ff00::/8",
      "2001:db8::/32",
      "2001:20::/28",
      "::1/128",
      "::/128",
      "100::/64",
      "64:ff9b::/96"
    ]
  end # def register
  
  public
  def filter(event)
    begin
      ip = event.get(@source)
      srcip = IPAddr.new(ip)
      validip = true
    rescue
      validip = false
    end
    if validip == true
      # ip version
      if srcip.ipv4?
        ipv = '4'
      elsif srcip.ipv6?
        ipv = '6'
      end
      event.set("#{@target_ipv}", ipv)

      # public ip check
      pubip = true
      @cidr.each do |p|
        cidr = IPAddr.new(p)
        if cidr.include?(srcip)
          pubip = false
        end
      end
      event.set("#{@target_pub_ip}", pubip)
    end

    # Tag event if invalid ip
    tag_invalid_ip(event) if validip == false
    # filter_matched should go in the last line of our successful code
    filter_matched(event) if validip == true
  end # def filter

  def tag_invalid_ip(event)
    @logger.debug? && @logger.warn("Invalid IP #{event.get(@source)}", :event => event)
    @tag_on_invalid_ip.each{|tag| event.tag(tag)}
  end # def tag_invalid_ip

end # class LogStash::Filters::PublicIp
